cmd/dp-observation-importer/main.go

```diff
@@ -2,76 +2,123 @@ package main
 
 import (
 	"context"
-	"net/http"
+	"fmt"
 	"os"
 	"os/signal"
+	"strconv"
 	"syscall"
-	"time"
 
+	"github.com/ONSdigital/dp-api-clients-go/dataset"
 	"github.com/ONSdigital/dp-graph/graph"
+	"github.com/ONSdigital/dp-healthcheck/healthcheck"
+	kafka "github.com/ONSdigital/dp-kafka"
 	"github.com/ONSdigital/dp-observation-importer/config"
 	"github.com/ONSdigital/dp-observation-importer/dimension"
 	"github.com/ONSdigital/dp-observation-importer/event"
+	"github.com/ONSdigital/dp-observation-importer/initialise"
 	"github.com/ONSdigital/dp-observation-importer/observation"
-	"github.com/ONSdigital/dp-reporter-client/reporter"
-	"github.com/ONSdigital/go-ns/handlers/healthcheck"
-	"github.com/ONSdigital/go-ns/kafka"
-	"github.com/ONSdigital/go-ns/log"
 	"github.com/ONSdigital/go-ns/server"
+	"github.com/ONSdigital/log.go/log"
 	"github.com/gorilla/mux"
 )
 
+var (
+	// BuildTime represents the time in which the service was built
+	BuildTime string
+	// GitCommit represents the commit (SHA-1) hash of the service that is running
+	GitCommit string
+	// Version represents the version of the service that is running
+	Version string
+)
+
 func main() {
 	log.Namespace = "dp-observation-importer"
+	ctx := context.Background()
 
-	cfg, err := config.Get()
-	checkForError(err)
+	signals := make(chan os.Signal, 1)
+	signal.Notify(signals, os.Interrupt, syscall.SIGTERM)
 
-	// Avoid logging the neo4j URL as it may contain a password
-	log.Debug("loaded config", log.Data{"config": cfg})
+	log.Event(ctx, "Starting observation importer", log.INFO)
 
-	// a channel used to signal a graceful exit is required.
-	errorChannel := make(chan error)
+	cfg, err := config.Get()
+	exitIfFatal(ctx, "failed to retrieve configuration", err)
```

Comment:
So I've created a run function but not sure what we have to gain out of it, James intends to update the log package to meet current logging standards for a fatal event; this would call os.Exit(1) and force the application to close and then the reasoning above breaks down unless we never use that event level unless in main.go. 

We also lose information with the current log lines (e.g. unique messages), this can be resolved in two ways but returning the message or wrap the error.

I feel like any further changes should be brought in as separate stories to update the applications as this firstly doesn't feel right and secondly is scope creep which I need to push back on; sorry

Reply:
Having the logging package shut down our app feels like a really bad idea - that should be down to our app to decide. Furthermore, how would graceful shutdown fit into that scenario? The logging library would need to know more about the app which is introducing a very tight coupling between the app and the library which feels wrong and like a code smell.

The reason I requested this change was so that any error that was deemed fatal would be propagated back to main and main would log and os.Exit - this gives us a single exit point. It sounds like there needs to be a discussion on this. I have a feeling I will be in a minority but then whats new? 🤷‍♂

Reply:
So I've read Mat Ryers [blog](https://pace.dev/blog/2020/02/12/why-you-shouldnt-use-func-main-in-golang-by-mat-ryer) and the main reason to create a run function (in my opinion) is to write tests which I do agree with. Writing these tests would increase the scope of this work and was not previously planned; I will however return all fatal errors back to main function from run

Next retro I will bring up that we need to write tests for our run functions and refactor some of the applications which don't currently split functionality out of main func into run.